### PR TITLE
Fix type hinting

### PR DESCRIPTION
**Pull request by mjpieters**
This PR fixes the type hinting issue #4.

- Use a `ParamSpec` to capture the wrapped function arguments and a
  separate `TypeVar` for the return type, so that the annotation can
  accurately wrap just the return value in a `Coroutine` annotation.
- Add explicit type hints for functions that use @overload (which
  are not compatible with ParamSpec).
- Add typing-extensions as a dependency (used for ParamSpec on older
  Python versions)
- Add a few simple tests to verify that type hints now are producing
  correct output.

Note that I switched to using `Coroutine[Any, Any, ...]` here, it's the more
accurate type hint.

### CODE REVIEW

**Comment by mjpieters** [comment]
The test failures here will be fixed once #7 is merged and I've rebased this on top of that PR.

**Comment by mjpieters** [comment]
Right, so the tests should now pass, because the mypy typechecks are _skipped_ on anything other than 3.10. That's because mypy refuses to acknowledge the backported ParamSpec object for now on anything other than 3.10; I think https://github.com/python/mypy/issues/12011 may be related there. This is a mypy bug, so I'm not taking the hints out again. :-)

File: aioshutil/__init__.py [hunk 1]
```diff
@@ -35,42 +53,183 @@
     "SameFileError",
 ]
 
-T = TypeVar("T", bound=Callable[..., Any])
+P = ParamSpec("P")
+R = TypeVar("R")
+
+if TYPE_CHECKING:  # pragma: no cover
+    # type hints for wrapped functions with overloads (which are incompatible
```
Comment (kumaraditya303) [inline aioshutil/__init__.py:60]: Oh, this is unfortunate that `ParamSpec` does not support overloaded functions.

File: tests/conftest.py [hunk 2]
```diff
@@ -4,4 +4,5 @@
 
 def pytest_collection_modifyitems(items):
     for item in items:
-        item.add_marker(pytest.mark.asyncio)
+        if isinstance(item, pytest.Function):
```
Comment (kumaraditya303) [inline tests/conftest.py:7]: Why is this needed?

Comment (mjpieters) [reply tests/conftest.py:7]: Because the mypy test cases (defined in the YAML file) break otherwise.

**Review by kumaraditya303 — CHANGES_REQUESTED** [review]
Thanks for the PR!

I have left some comments and there is a conflict, after you have fixed those then this is ready to go in.

File: aioshutil/__init__.py [hunk 3]
```diff
@@ -35,42 +53,183 @@
...
+    async def copy(
+        src: BytesPath, dst: BytesPath, *, follow_symlinks: bool = ...
+    ) -> _PathReturn:
+        ...
+
+    async def copy(src, dst, *, follow_symlinks=...):
+        ...
+
+    @overload
+    def copy2(
```
Comment (kumaraditya303) [inline aioshutil/__init__.py:88]: This should also be `async`, no?

**Comment by mjpieters** [comment]
Bah, humbug. I'll have to fix the PR push. 

I was trying to do this via GitHub codespaces on a mobile phone as I am travelling.

**Review by kumaraditya303 — APPROVED** [review]
LGTM Thanks!
